src/Persistence/Array_/Query.php

```diff
@@ -2,48 +2,188 @@
+            array_multisort(...$args);
+
+            // put data back in generator
+            $this->iterator = new \ArrayIterator(array_pop($args));
+        }
+    }
+
+    protected function initLimit(): void
+    {
+        if ($args = $this->getLimitArgs()) {
+            [$limit, $offset] = $args;
+
+            $this->iterator = new \LimitIterator($this->iterator, $offset, $limit);
+        }
+    }
 
     /**
-     * Iterator constructor.
+     * Counts number of rows and replaces our generator with just a single number.
+     *
+     * @return $this
      */
-    public function __construct(array $data)
+    protected function initCount($alias = null): void
     {
-        $this->generator = new \ArrayIterator($data);
+        // @todo: kept for BC, inconstent results with SQL count!
+        $this->initLimit();
+
+        $alias = $alias ?? 'count';
+
+        $this->iterator = new \ArrayIterator([[$alias => iterator_count($this->iterator)]]);
     }
 
     /**
-     * Applies FilterIterator making sure that values of $field equal to $value.
-     *
-     * @param string $field
-     * @param string $value
+     * Checks if iterator has any rows.
      *
      * @return $this
      */
-    public function filter(Model\Scope\AbstractScope $condition)
+    protected function initExists(): void
     {
-        if (!$condition->isEmpty()) {
-            $this->generator = new \CallbackFilterIterator($this->generator, function ($row) use ($condition) {
-                return $this->match($row, $condition);
-            });
+        $this->iterator = new \ArrayIterator([[$this->iterator->valid() ? 1 : 0]]);
+    }
+
+    protected function initField($fieldName, string $alias = null): void
+    {
+        if (!$fieldName) {
+            throw new Exception('Field query requires field name');
+        }
+
+        $field = $fieldName;
+        if (!is_string($fieldName)) {
+            $field = $this->model->getField($fieldName);
+            $fieldName = $field->short_name;
         }
 
-        return $this;
+        $this->initSelect([$fieldName]);
+
+        // get first record
+        if ($row = $this->getRow()) {
+            if ($alias && array_key_exists($field, $row)) {
+                $row[$alias] = $row[$field];
+                unset($row[$field]);
+            }
+        }
+
+        $this->iterator = new \ArrayIterator([[$row]]);
+    }
+
+    protected function doExecute(): iterable
+    {
+        return $this->iterator;
+    }
+
+    /**
+     * Return all data inside array.
+     */
+    public function get(): array
+    {
+        return iterator_to_array($this->iterator, true);
+    }
+
+    /**
+     * Return one row of data.
+     */
+    public function getRow(): ?array
+    {
+        $row = $this->iterator->current();
+
+        $this->iterator->next();
+
+        return $row;
+    }
+
+    /**
+     * Return one value from one row of data.
+     *
+     * @return mixed
+     */
+    public function getOne()
```

Comment:
like in dsql, it should have the string `?string` return type

Reply:
`?string` clashes with current setup

https://github.com/atk4/data/blob/e64b392d057243510e62be797ff405bc598ff06a/tests/PersistentArrayTest.php#L333

Reply:
Hmm, quite important, please look at this, there is something wrong, we need same result level across all persistences.

Reply:
Fully agree results of the query should be uniform across the persistences.
But I would split this in 2 stages:
1. finlize and merge the reworked classes in fully BC way
2. apply limitiations as deemed necessary which may introduce BC-breaks. It will be easy once 1. is in place.